### PR TITLE
p^d L2L and M2M

### PR DESCRIPTION
**Pull request by isuruf**
Mirror of https://gitlab.tiker.net/inducer/sumpy/-/merge_requests/125 for review

cc @mattwala

### CODE REVIEW

File: sumpy/expansion/local.py [hunk 1]
```diff
@@ -127,90 +127,19 @@ def coefficients_from_source(self, avec, bvec, rscale, sac=None):
...
-        # }}}
+        from sumpy.tools import mi_power, mi_factorial
 
-        return local_sum[0]
+        return sum(
+            coeff
+            * mi_power(bvec, mi, evaluate=False)
+            / mi_factorial(mi)
+            for coeff, mi in zip(
+                    evaluated_coeffs, self.get_full_coefficient_identifiers()))
```
Comment (inducer) [inline sumpy/expansion/local.py:142]: Could you explain the reasoning behind bringing this back?

File: sumpy/expansion/multipole.py [hunk 2]
```diff
@@ -182,54 +183,112 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+                            contrib /= mi_factorial((n-k,))
+                            contrib *= \
+                                sym.UnevaluatedExpr(dvec[idim]/tgt_rscale)**(n-k)
+
+                        temp[i] += contrib
+
+                dim_coeffs_to_translate = temp[:]
+
+            for i, mi in enumerate(src_expansion.get_full_coefficient_identifiers()):
+                result[tgt_mi_to_index[mi]] += dim_coeffs_to_translate[i]
```
Comment (inducer) [inline sumpy/expansion/multipole.py:252]: As above, I have no idea what this is even supposed to do.

**Review by inducer — COMMENTED** [review]
Thanks for working on this. I'll be honest. I don't understand any part of this PR. Could you try and explain it better?

File: sumpy/expansion/local.py [hunk 3]
```diff
@@ -301,9 +230,68 @@ def calc_max_mi(mis):
...
+        tgt_mis_all = \
+            self.expansion_terms_wrangler.get_coefficient_identifiers()
+        tgt_mi_to_index = dict((mi, i) for i, mi in enumerate(tgt_mis_all))
+
+        tgt_split = self.expansion_terms_wrangler._get_coeff_identifier_split()
+
+        p = max(sum(mi) for mi in src_mis)
+        result = [0] * len(tgt_mis_all)
+
+        # O(1) iterations
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: It is O(1) even for full. Explained in detail in a code comment.

File: sumpy/expansion/local.py (hunk 1 above)
Comment (isuruf) [reply sumpy/expansion/local.py:142]: Previous complicated code was necessary for a better L2L, but it didn't improve L2P at all. Now that there's a separate algorithm for L2L, I've reverted back to the simpler version.

File: sumpy/expansion/__init__.py [hunk 4]
```diff
@@ -208,6 +209,66 @@ def copy(self, **kwargs):
...
+        which is orthogonal to one of the main axis.
+
+        If this is an instance of LinearPDEBasedExpansionTermsWrangler,
+        then the number of sets will be :math:`O(1)`.
+
+        Returns an object of type List[Tuple[int, List[Tuple[int]]]] where
+        each element in the outer list represents the hyperplane. Each element
+        is a 2-tuple where the first element in the tuple is the axis number d
+        to which the hyperplane is orthogonal to. Second element in the tuple
+        is a list of multi-indices in the hyperplane.
```
Comment (inducer) [inline sumpy/expansion/__init__.py:228]:
```suggestion
        In the returned object ``[(axis, [mi_1, mi2, ...]), ...]``,
        each element in the outer list represents a hyperplane. Each element
        is a 2-tuple where the first element in the tuple is the axis number *axis*
        to which the hyperplane is orthogonal. The second element in the tuple
        is a list of multi-indices in the hyperplane corresponding to the stored
        coefficients.
        
        E.g. for Laplace 3D order 4, the result could be::
        
            [
            (2, [(0, 0, 0), (1, 0, 0), (2, 0, 0), (3, 0, 0), (0, 1, 0), (1, 1, 0), (2, 1, 0), (0, 2, 0), (1, 2, 0), (0, 3, 0)]),
            (2, [(0, 0, 1), (1, 0, 1), (2, 0, 1), (0, 1, 1), (1, 1, 1), (0, 2, 1)]),
            ]
```

File: sumpy/expansion/__init__.py [hunk 5]
```diff
@@ -208,6 +209,66 @@ def copy(self, **kwargs):
...
         return type(self)(**new_kwargs)
 
+    @memoize_method
+    def _get_coeff_identifier_split(self) -> List[Tuple[int, List[Tuple[int]]]]:
+        """
+        This splits the coefficients into :math:`O(p)` number of disjoint sets
+        so that for each set, all the identifiers have the form,
+        :math:`(m_1, m_2, ..., m_{j-1}, c, m_{j+1}, ... , m_{dim})`
+        where :math:`c` is a constant. Geometrically, each set is a hyperplane
+        which is orthogonal to one of the main axis.
```
Comment (inducer) [inline sumpy/expansion/__init__.py:219]:
```suggestion
        where :math:`c` is a constant. Geometrically, each set is an axis-aligned
        hyperplane.
```

**Review by inducer — COMMENTED** [review]
(Half-done review during meeting)

File: sumpy/expansion/multipole.py [hunk 6]
```diff
@@ -166,67 +166,129 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
         # and can be rewritten as follows.
         #
-        # Let $T_{m, n} = \sum_{i\le m} A_{i, n} d_x^i \binom{m}{i}$.
+        # Let $Y_{m, n} = \sum_{i\le m} C_{i, n} d_x^i \binom{m}{i}$.
         #
-        # Then, $B_{m, n} = \sum_{j\le n} T_{m, j} d_y^j \binom{n}{j}$.
+        # Then, $T_{m, n} = \sum_{j\le n} Y_{m, j} d_y^j \binom{n}{j}$.
         #
-        # $T_{m, n}$ are $p^2$ number of temporary variables that are
+        # $Y_{m, n}$ are $p^2$ number of temporary variables that are
```
Comment (inducer) [inline sumpy/expansion/multipole.py:180]:
```suggestion
        # $Y_{m, n}$ are $p^2$ temporary variables that are
```

File: sumpy/expansion/multipole.py [hunk 7]
```diff
@@ -166,67 +166,129 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+                        new_mi[d] = mi_i
+                        input_mis_per_dim.append(tuple(new_mi))
+
+                    for input_mi in input_mis_per_dim:
+                        contrib = C[tgt_mi_to_index[input_mi]]
+                        for idim in range(self.dim):
+                            n = tgt_mi[idim]
+                            k = input_mi[idim]
+                            assert n >= k
+                            contrib /= mi_factorial((n-k,))
```
Comment (inducer) [inline sumpy/expansion/multipole.py:243]: This is a scalar factorial, right? Why involve `mi_factorial` for this?

File: sumpy/expansion/multipole.py [hunk 8]
```diff
@@ -166,67 +166,129 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
 
-        mi_to_index = src_mi_to_index
-        for d in range(self.dim):
+        # For this algorithm, we need the hyperplanes that are parallel
+        # to each other grouped.
+        non_zero_coeffs_grouped_by_normal_dir = [[] for d in range(self.dim)]
+        for d, tgt_mi in tgt_split:
+            non_zero_coeffs_grouped_by_normal_dir[d] += tgt_mi
+
+        for normal_dir in set(d for d, _ in tgt_split):
```
Comment (inducer) [inline sumpy/expansion/multipole.py:206]: These two loops could probably be fused, avoiding the need for the `non_zero_coeffs_grouped_by_normal_dir` temporary.

File: sumpy/expansion/multipole.py [hunk 9]
```diff
@@ -166,67 +166,129 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+                    # Calling this input_mis instead of src_mis because we
+                    # converted the source coefficients to target coefficient
+                    # indices beforehand.
+                    input_mis_per_dim = []
+                    for mi_i in range(tgt_mi[d]+1):
+                        new_mi = list(tgt_mi)
+                        new_mi[d] = mi_i
+                        input_mis_per_dim.append(tuple(new_mi))
+
+                    for input_mi in input_mis_per_dim:
```
Comment (inducer) [inline sumpy/expansion/multipole.py:237]: These two loops could be fused, avoiding the need for `input_mis_per_dim`.

File: sumpy/expansion/multipole.py [hunk 10]
```diff
@@ -166,67 +166,129 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+                src_idx = src_mi_to_index[mi]
+                tgt_idx = tgt_mi_to_index[mi]
+                C[tgt_idx] = src_coeff_exprs[src_idx] * \
+                        sym.UnevaluatedExpr(src_rscale/tgt_rscale)**sum(mi)
+
+            # Use the normal_dir as the last dimension to vary
+            dims = list(range(normal_dir)) + \
+                   list(range(normal_dir+1, self.dim)) + [normal_dir]
+            for d in dims:
+                Y = [0] * len(self.get_full_coefficient_identifiers())  # noqa: N806
```
Comment (inducer) [inline sumpy/expansion/multipole.py:224]:
```suggestion
                # We build the full target multipole and then compress it, below.
                Y = [0] * len(self.get_full_coefficient_identifiers())  # noqa: N806
```

**Review by inducer — COMMENTED** [review]
Thanks for working on this. This was still a super painful, multi-hour slog for me. :/

File: sumpy/expansion/__init__.py [hunk 11]
```diff
@@ -208,6 +209,74 @@ def copy(self, **kwargs):
...
+        if not hyperplanes:
+            d = self.dim - 1
+            for const in range(self.order + 1):
+                hyperplanes.append((d, const))
+
+        res = []
+        seen_mis = set()
+        for d, const in hyperplanes:
+            coeffs_in_hyperplane = []
+            for mi in self.get_coefficient_identifiers():
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/__init__.py:268]: It's the other way around. After finding the stored coefficient identifiers in `self.get_coefficient_identifiers`, this function divides them into hyperplanes.

File: sumpy/expansion/__init__.py [hunk 12]
```diff
@@ -208,6 +209,74 @@ def copy(self, **kwargs):
...
+                # Check if the multi-index is in this hyperplane and
+                # if it is not in any of the hyperplanes we saw before
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/__init__.py:270]: There's no sorting assumption. The only requirement is that if a multi-index is in the intersection of two hyperplanes, it should be put into one and only one hyperplane.

File: sumpy/expansion/multipole.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/multipole.py]: Done in 6e94722

File: sumpy/expansion/__init__.py [hunk 13]
```diff
@@ -208,6 +209,76 @@ def copy(self, **kwargs):
 
         return type(self)(**new_kwargs)
 
+    @memoize_method
+    def _get_coeff_identifier_split(self) -> List[Tuple[int, List[Tuple[int]]]]:
```
Comment (inducer) [inline sumpy/expansion/__init__.py:213]:
```suggestion
    def _split_coeffs_into_hyperplanes(self) -> List[Tuple[int, List[Tuple[int]]]]:
```

File: sumpy/expansion/local.py [hunk 14]
```diff
@@ -298,9 +229,107 @@ def calc_max_mi(mis):
...
+        src_mis = \
+            src_expansion.expansion_terms_wrangler.get_full_coefficient_identifiers()
+
+        src_mi_to_index = {mi: i for i, mi in enumerate(src_mis)}
+
+        tgt_mis = \
+            self.expansion_terms_wrangler.get_coefficient_identifiers()
+        tgt_mi_to_index = {mi: i for i, mi in enumerate(tgt_mis)}
+
+        tgt_split = self.expansion_terms_wrangler._get_coeff_identifier_split()
```
Comment (inducer) [inline sumpy/expansion/local.py:252]: Rename `tgt_split` to something with hyperplanes?

File: sumpy/expansion/multipole.py [hunk 15]
```diff
@@ -166,67 +167,205 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
-        dim_coeffs_to_translate = src_coeff_exprs
+        # one dimension at a time. If the coefficient-identifying multi-indices
+        # in the source expansion have the form (0, m) and (n, 0), where m>=0, n>=1,
+        # then we calculate the output from (0, m) with the second
+        # dimension as the fastest varying dimension and then calculate
+        # the output from (n, 0) with the first dimension as the fastest
+        # varying dimension.
+
+        tgt_split = \
+            self.expansion_terms_wrangler._get_coeff_identifier_split()
```
Comment (inducer) [inline sumpy/expansion/multipole.py:264]: Rename `tgt_split` to something with hyperplanes.

File: sumpy/expansion/multipole.py [hunk 16]
```diff
@@ -166,67 +167,205 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+        # hyperplane.
+        #
+        #    0              0            0
+        #
+        #    0 0       ->   0┌0      ->  0 T
+        #                    ││
+        #    0 0 0          0│0 0        0 T T
+        #                    ││ │
+        #    0 C-C-C        0└Y Y Y      0 T T T
+        #      └───┘
```
Comment (inducer) [inline sumpy/expansion/multipole.py:253]: Arrows? Or maybe just judiciously placed `>` signs?

**Review by inducer — COMMENTED** [review]
Thanks for bearing with me here. A few simple things, and this is ready from my perspective.

**Comment by isuruf** [comment]
What do you think about 48c069a? I'm not a huge fan, but that's the best we can do I think.

**Comment by inducer** [comment]
Thanks for trying! I think it's an improvement, and good enough for me.

**Comment by inducer** [comment]
Is this ready for another look?

**Comment by isuruf** [comment]
> Is this ready for another look?

It is now.

**Comment by isuruf** [comment]
@inducer, I fixed the formulae in the comments for L2L.

File: sumpy/expansion/multipole.py [hunk 17]
```diff
@@ -166,67 +167,206 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+        #
+        #  C              C             0
+        #  C 0            C 0           0 0
+        #  C 0 0       =  C 0 0      +  0 0 0
+        #  C 0 0 0        C 0 0 0       0 0 0 0
+        #  C C C C C      C 0 0 0 0     0 C C C C
+        #
+        # The calculations done when translating first hyperplane of the
+        # source coefficients (C) to target coefficients (T) are shown
+        # below in the graph. Each connection represents a O(1) calculation.
```
Comment (inducer) [inline sumpy/expansion/multipole.py:199]:
```suggestion
        # The calculations done when naively translating first hyperplane of the
        # source coefficients (C) to target coefficients (T) are shown
        # below in the graph. Each connection represents a O(1) calculation,
        # and the arrows go "up and to the right".
```

File: sumpy/expansion/multipole.py [hunk 18]
```diff
@@ -166,67 +167,206 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+                            sym.UnevaluatedExpr(src_rscale/tgt_rscale)**sum(mi)
+
+            if all(coeff == 0 for coeff in cur_dim_input_coeffs):
+                continue
+
+            # }}}
+
+            # {{{ translation
+
+            # Therefore, we use the orthogonal axis as the last dimension to vary
```
Comment (inducer) [inline sumpy/expansion/multipole.py:298]: > Therefore

What is this referring to?

File: sumpy/expansion/multipole.py [hunk 19]
```diff
@@ -166,67 +167,206 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+            for d, mis in tgt_hyperplanes:
+                # Only consider hyperplanes perpendicular to *axis*.
+                if d != axis:
+                    continue
+                for mi in mis:
+                    # When target order is higher than source order, we assume
+                    # that the higher order source coefficients were zero.
+                    if mi not in src_mi_to_index:
+                        continue
+                    src_idx = src_mi_to_index[mi]
```
Comment (inducer) [inline sumpy/expansion/multipole.py:286]:
```suggestion
                    
                    src_idx = src_mi_to_index[mi]
```

**Review by inducer — COMMENTED** [review]
Thanks! A few more readability/follow-ability concerns, and this should be good to go.

**Comment by inducer** [comment]
:tada: In it goes, and looks great. Thanks for your patience with all my readability concerns!

**Pull request merged**